solidity/dashboard/src/components/Rewards.jsx

```diff
@@ -0,0 +1,349 @@
+    rewardsService.fetchWithdrawalHistory,
+    []
+  )
+  const { data: withdrawals } = withdrawalHistoryState
+
+  // see more/less button state
+  const [showAll, setShowAll] = useState(false)
+
+  // filter dropdown
+  const [rewardFilter, setRewardFilter] = useState({})
+
+  const withdrawnEventCallback = (latestEvent) => {
+    const {
+      transactionHash,
+      blockNumber,
+      returnValues: { groupIndex, amount, beneficiary, operator },
+    } = latestEvent
+
+    if (!isSameEthAddress(yourAddress, beneficiary)) {
+      return
+    }
+
+    updateRewards(latestEvent)
+    keepRandomBeaconOperatorContract.methods
+      .getGroupPublicKey(groupIndex)
+      .call()
+      .then((groupPublicKey) => {
+        const withdrawal = {
+          blockNumber,
+          groupPublicKey,
+          transactionHash,
+          reward: web3Utils.fromWei(amount, "ether"),
+          operatorAddress: operator,
+          status: REWARD_STATUS.WITHDRAWN,
+          groupStatus: SIGNING_GROUP_STATUS.COMPLETED,
+        }
+        updateWithdrawalHistoryData([withdrawal, ...withdrawals])
+      })
+  }
+
+  // subscribe to `GroupMemberRewardsWithdrawn` event
+  useSubscribeToContractEvent(
+    OPERATOR_CONTRACT_NAME,
+    "GroupMemberRewardsWithdrawn",
+    withdrawnEventCallback
+  )
+
+  const updateRewards = (latestEvent) => {
+    const {
+      returnValues: { groupIndex, operator, amount },
+    } = latestEvent
+    const { indexInArray } = findIndexAndObject(
+      "groupIndex",
+      groupIndex,
+      rewards,
+      (object) => isSameRewardRecord(object, groupIndex, operator)
+    )
+
+    if (indexInArray === null) {
+      return
+    }
+
+    const updateTotalRewardsBalance = sub(
+      web3Utils.toWei(totalRewardsBalance, "ether"),
+      amount
+    )
+    const updatedRewards = [...rewards]
+    updatedRewards.splice(indexInArray, 1)
+
+    updateData([
+      updatedRewards,
+      web3Utils.fromWei(updateTotalRewardsBalance, "ether"),
+    ])
+  }
+
+  const withdrawReward = async (
+    operatorAddress,
+    groupIndex,
+    onTransactionHashCallback
+  ) => {
+    try {
+      updateRewardStatus(PENDING_STATUS, groupIndex, operatorAddress)
+      await rewardsService.withdrawRewardFromGroup(
+        web3Context,
+        { operatorAddress, groupIndex },
+        onTransactionHashCallback
+      )
+      showMessage({
+        type: messageType.SUCCESS,
+        title: "Success",
+        content: "Withdrawal successfully completed",
+      })
+    } catch (error) {
+      showMessage({
+        type: messageType.ERROR,
+        title: "Withdrawal action has failed ",
+        content: error.message,
+      })
+      throw error
+    }
+  }
+
+  const updateRewardStatus = (status, groupIndex, operator) => {
+    const { indexInArray } = findIndexAndObject(
+      "groupIndex",
+      groupIndex,
+      rewards,
+      (object) => isSameRewardRecord(object, groupIndex, operator)
+    )
+    if (indexInArray === null) {
```

Comment:
This doesn't protect against `undefined`. Are we good here?

Reply:
^ comment above